### PR TITLE
Don't fill notes that overlap slightly

### PR DESCRIPTION
**Pull request by allejok96**
If a note overlaps the next by a hair, pressing shift+F will fill it all the way across the other note. That's annoying.

This PR fixes that by only filling overlapping notes if the overlap by more than 50%

bild

### CODE REVIEW

**Comment by sakertooth** [comment]
How do other DAWs fill notes? Trying to make sure that this is the expected behavior.

File: src/gui/editors/PianoRoll.cpp [hunk 1]
```diff
@@ -766,6 +770,79 @@ void PianoRoll::fitNoteLengths(bool fill)
 }
 
 
+
+bool fillGapsBetweenNotesInVector(
```
Comment (messmerd) [inline src/gui/editors/PianoRoll.cpp:774]:
```suggestion
static bool fillGapsBetweenNotesInVector(
```
This could be static since it's only used in this translation unit.

File: src/gui/editors/PianoRoll.cpp [hunk 2]
```diff
@@ -711,8 +711,12 @@ void PianoRoll::glueNotes()
 	}
 }
 
-void PianoRoll::fitNoteLengths(bool fill)
+void PianoRoll::cutOverlappingNotes()
 {
+	// TODO remove the code for fill=true
+	// it has been replaced by fillGapsBetweenNotes()
+	bool fill = false;
+
```
Comment (regulus79) [inline src/gui/editors/PianoRoll.cpp:719]: Would it make sense to also remove the old code as part of this PR?

Comment (allejok96) [reply src/gui/editors/PianoRoll.cpp:717]: I did that initially, but it created such a messy and hard-to-review diff for this PR so I think it better be done in a later PR.

**Comment by allejok96** [comment]
Merge conflicts resolved. Sorry for the long wait.

> How do other DAWs fill notes? Trying to make sure that this is the expected behavior.

I've only ever used LMMS. I don't know if this is the *expected* behavior but I think the current one is *unexpected*. If a note overlaps the next by 1/192 it will be filled across it. It's such a tiny thing you probably won't even see it until you try to use Fill. This is similar to how clips become a bar longer if the last note stretches over the bar line by 1/192. It's one of those things that can be very hard and frustrating for new users to figure out how and why.

**Comment by rubiefawn** [comment]
> How do other DAWs fill notes?

Ableton Live user reporting in: Ableton does have a similar piano roll "legato" function. However, it shouldn't be considered for reasons I'll demonstrate. Here's an approximation of the example MIDI shown above:

image

Here it is after applying legato:

image

This sometime shortens notes, but it isn't clear how this function decides whether it is lengthening or shortening a note, especially when they are overlapping. This is the in-program tooltip for legato (which is more descriptive than the manual page):

image

> Trying to make sure that this is the expected behavior.

In conclusion, Ableton's legato function is not really the same as LMMS's fill function, since fill does not shorten notes, and it also works predictably on polyphonic selections. So there's at least one commercial DAW out there where this PR doesn't violate user expectations. 🤷‍♀️

**Comment by allejok96** [comment]
I really like that logic! Though since we have a separate Cut overlaps function we can let the shortening be optional.

But Cut overlaps could need some smarter logic too... Ardour has a similar issue with it's legato function:

bild
bild

**Comment by CorruptVoidSoul** [comment]
I tested this and as of now, it works as you intended.
Funny how that made me discover a feature I never used but is actually really neat.

File: src/gui/editors/PianoRoll.cpp (hunk 2 above)
Comment (messmerd) [reply src/gui/editors/PianoRoll.cpp:711]: It doesn't look like it would be that difficult to review if you removed it, so I'd say go ahead. You know what they say, "there is nothing more permanent than a temporary solution"

File: src/gui/editors/PianoRoll.cpp [hunk 3]
```diff
@@ -758,6 +762,79 @@ void PianoRoll::fitNoteLengths(bool fill)
...
+	m_midiClip->addJournalCheckPoint();
+
+	// Make sure notes are sorted by start position
+	m_midiClip->rearrangeAllNotes();
+
+	bool hasSelection = !getSelectedNotes().empty();
+
+	fillGapsBetweenNotesInVector(
+		m_midiClip->notes(),
+		/*onlyEditSelected*/ hasSelection);
```
Comment (messmerd) [inline src/gui/editors/PianoRoll.cpp:833]: If this returns false (meaning nothing changed), should you call `m_midiClip->restoreJournallingState()`?

Otherwise - if I'm understanding how journalling works correctly - there would be a journalling checkpoint that doesn't change anything, and I'm not sure if that's desirable.
